File: 16-ArithGeo/16.hs
--Asumming no zeros, detect if an array of numbers follows a artihmetic or geometric pattern

arithGeo :: (Eq a, Fractional a) => [a] -> Maybe String
arithGeo list@(x:y:z:rest)
  | x + (z-y) == y = arith list
  | otherwise      = geo list
    where arith (x:y:z:rest)
            | x + (z-y) == y   = if null rest then Just "Arithmetic" else arith (y:z:rest)
            | otherwise        = Nothing
          geo (x:y:z:rest)
            | x * (z/y) == y   = if null rest then Just "Geometric" else geo (y:z:rest)
            | otherwise        = Nothing

arithGeo arr = Nothing

main = do
  print $ arithGeo [2,4,6,8]   == Just "Arithmetic"
  print $ arithGeo [2,4,6,10]  == Nothing
  print $ arithGeo [3,9,27,81] == Just "Geometric"
  print $ arithGeo [1,4,16,20] == Nothing
  print $ arithGeo [1,2,3,4.5] == Nothing 

File: 16-ArithGeo/16.js
// Asumming no zeros, detect if an array of numbers follows a artihmetic or geometric pattern

function ArithGeo(arr){
  return (arr[0] + (arr[2]-arr[1]) === arr[1]) ? arith(arr) : geo(arr)

  function arith(arr){
    return (arr[0] + (arr[2]-arr[1]) === arr[1] && (arr.length === 3 || arith(arr.slice(1)) === "Arithmetic")) ? "Arithmetic" : -1
  }

  function geo(arr){
    return (arr[0] * (arr[2]/arr[1]) === arr[1] && (arr.length === 3 || geo(arr.slice(1)) === "Geometric")) ?  "Geometric" : -1
  }
}

console.log(ArithGeo([2,4,6,8])   === "Arithmetic")
console.log(ArithGeo([2,4,6,10])  === -1)
console.log(ArithGeo([3,9,27]) === "Geometric")
console.log(ArithGeo([1,4,16,20])  === -1)
console.log(ArithGeo([1,2,3,4.5])  === -1)